{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "crude-transparency",
   "metadata": {},
   "source": [
    "## References"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "indian-narrative",
   "metadata": {},
   "source": [
    "Tutorial: https://docs.scvi-tools.org/en/stable/user_guide/notebooks/MultiVI_tutorial.html <br>\n",
    "Paper: https://www.biorxiv.org/content/10.1101/2021.08.20.457057v2"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "structural-vehicle",
   "metadata": {},
   "source": [
    "## Dataset to prepare"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "heated-scene",
   "metadata": {},
   "source": [
    "### 1) RNA (scnRNA + Multiome-RNA)\n",
    "* Read in data: post-CellBender, filtered as the previous HCA object, cell-type annotated\n",
    "* Subset scnRNA: barcode x gene -> **`adata_rna.h5ad`**\n",
    "* Subset MultiomeRNA: barcode x gene\n",
    "\n",
    "### 2) ATAC (snATAC + Multiome-ATAC)\n",
    "* Read in data: post-cellatac and filtered peaks and nuclei, `6reg-v2_ATACs_filtered.h5ad`\n",
    "* Subset snATAC: barcode x peak -> **`adata_atac.h5ad`**\n",
    "* Subset MultiomeATAC: barcode x peak\n",
    "\n",
    "### 3) Concatenate Multiome RNA+ATAC\n",
    "barcode x (gene+peak) -> **`adata_paired.h5ad`**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "robust-count",
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "import anndata\n",
    "import scipy"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "civic-butterfly",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "anndata             0.7.6\n",
       "numpy               1.20.3\n",
       "pandas              1.3.2\n",
       "scanpy              1.8.1\n",
       "scipy               1.7.1\n",
       "session_info        1.0.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                         8.3.2\n",
       "backcall                    0.2.0\n",
       "beta_ufunc                  NA\n",
       "binom_ufunc                 NA\n",
       "cffi                        1.14.6\n",
       "colorama                    0.4.4\n",
       "cycler                      0.10.0\n",
       "cython_runtime              NA\n",
       "dateutil                    2.8.2\n",
       "debugpy                     1.4.1\n",
       "decorator                   5.0.9\n",
       "defusedxml                  0.7.1\n",
       "entrypoints                 0.3\n",
       "google                      NA\n",
       "h5py                        3.4.0\n",
       "ipykernel                   6.3.1\n",
       "ipython_genutils            0.2.0\n",
       "jedi                        0.18.0\n",
       "joblib                      1.0.1\n",
       "kiwisolver                  1.3.2\n",
       "llvmlite                    0.37.0\n",
       "matplotlib                  3.4.3\n",
       "matplotlib_inline           NA\n",
       "mpl_toolkits                NA\n",
       "natsort                     7.1.1\n",
       "nbinom_ufunc                NA\n",
       "numba                       0.54.0\n",
       "numexpr                     2.7.3\n",
       "packaging                   21.0\n",
       "parso                       0.8.2\n",
       "pexpect                     4.8.0\n",
       "pickleshare                 0.7.5\n",
       "pkg_resources               NA\n",
       "prompt_toolkit              3.0.20\n",
       "ptyprocess                  0.7.0\n",
       "pydev_ipython               NA\n",
       "pydevconsole                NA\n",
       "pydevd                      2.4.1\n",
       "pydevd_concurrency_analyser NA\n",
       "pydevd_file_utils           NA\n",
       "pydevd_plugins              NA\n",
       "pydevd_tracing              NA\n",
       "pygments                    2.10.0\n",
       "pyparsing                   2.4.7\n",
       "pytz                        2021.1\n",
       "sinfo                       0.3.4\n",
       "six                         1.16.0\n",
       "sklearn                     0.24.2\n",
       "storemagic                  NA\n",
       "tables                      3.6.1\n",
       "tornado                     6.1\n",
       "traitlets                   5.1.0\n",
       "typing_extensions           NA\n",
       "wcwidth                     0.2.5\n",
       "yaml                        5.4.1\n",
       "zmq                         22.2.1\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             7.27.0\n",
       "jupyter_client      7.0.2\n",
       "jupyter_core        4.7.1\n",
       "notebook            6.4.3\n",
       "-----\n",
       "Python 3.8.10 | packaged by conda-forge | (default, May 11 2021, 07:01:05) [GCC 9.3.0]\n",
       "Linux-4.15.0-156-generic-x86_64-with-glibc2.10\n",
       "-----\n",
       "Session information updated at 2021-09-21 13:35\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import session_info\n",
    "session_info.show()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "cosmetic-destiny",
   "metadata": {},
   "outputs": [],
   "source": [
    "def downsize_cell(adata, groupby, n_down, seed):\n",
    "   \n",
    "    groups = adata.obs[groupby].cat.categories\n",
    "    for group in groups:\n",
    "        adata_group = adata[adata.obs[groupby].isin([group])]\n",
    "        if len(adata_group)>n_down:\n",
    "            np.random.seed(seed)\n",
    "            rdcell = np.random.choice(adata_group.shape[0], size=n_down, replace=False)\n",
    "            if group == groups[0]:\n",
    "                down = adata_group[rdcell, :]\n",
    "            else:\n",
    "                down = down.concatenate(adata_group[rdcell, :], index_unique=None)\n",
    "        else:\n",
    "            if group == groups[0]:\n",
    "                down = adata_group\n",
    "            else:\n",
    "                down = down.concatenate(adata_group, index_unique=None)\n",
    "        del adata_group\n",
    "    \n",
    "    down.obs[groupby] = down.obs[groupby].astype('category')\n",
    "    return down"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "waiting-rapid",
   "metadata": {},
   "source": [
    "## Read in data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "fabulous-secondary",
   "metadata": {},
   "source": [
    "**RNA**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "shaped-connection",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 473643 × 31915\n",
       "    obs: 'sangerID', 'modality', 'donor', 'age_group', 'region', 'cell_or_nuclei', 'gender', 'type', 'cell_states', 'modality_fine', 'donor_cellnuc'\n",
       "    var: 'gene_ids-0', 'gene_ids-1', 'modality'"
      ]
     },
     "execution_count": 4,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna=sc.read_h5ad('/nfs/team205/kk18/HeartAtlas/anndata/adult-heart_scnRNA-MultiRNA_raw_210720.h5ad')\n",
    "rna.obs['modality_fine']=rna.obs['modality'].copy()\n",
    "\n",
    "# add intra-batchkey: 'donor_cellnuc'\n",
    "rna.obs['donor_cellnuc']=rna.obs['donor'].astype('str')+'_'+rna.obs['cell_or_nuclei'].astype('str')\n",
    "\n",
    "# add modality to .var\n",
    "rna.var['modality']='Gene Expression'\n",
    "\n",
    "rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "technological-topic",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1., 1., 1., 1., 4., 1., 1., 1., 1., 1.], dtype=float32)"
      ]
     },
     "execution_count": 5,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "rna.X.data[:10]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "ethical-communist",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "scnRNA      443004\n",
       "Multiome     30639\n",
       "Name: modality, dtype: int64"
      ]
     },
     "execution_count": 6,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# set 'modality' in .obs as 'scnRNA' and 'Multiome'\n",
    "rna.obs.replace({'modality':{\n",
    "    '3prime_v3':'scnRNA',\n",
    "    '3prime_v2':'scnRNA',\n",
    "}},inplace=True)\n",
    "\n",
    "rna.obs['modality'].value_counts()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "administrative-latter",
   "metadata": {},
   "source": [
    "**ATAC**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "moderate-charter",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 79208 × 102627\n",
       "    obs: 'cellatac_clusters', 'cellatac_code', 'sangerID', 'dataset', 'donor', 'Region', 'n_genes_by_counts', 'log1p_n_genes_by_counts', 'total_counts', 'log1p_total_counts', 'pct_counts_in_top_50_genes', 'pct_counts_in_top_100_genes', 'pct_counts_in_top_200_genes', 'pct_counts_in_top_500_genes', 'barcode', 'oribarcode', 'cell_or_nuclei', 'donor_cellnuc'\n",
       "    var: 'peak_width', 'exon', 'gene', 'promoter', 'annotation', 'gene_name', 'gene_id', 'tss_distance', 'ENCODE_blacklist', 'n_cells_by_counts', 'mean_counts', 'log1p_mean_counts', 'pct_dropout_by_counts', 'total_counts', 'log1p_total_counts', 'modality'\n",
       "    layers: 'binary_raw'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac = sc.read_h5ad('/nfs/team205/kk18/data/6region_v2/ATAC/6reg-v2_ATACs_filtered.h5ad')\n",
    "\n",
    "# add intra-batchkey: 'donor_cellnuc'\n",
    "atac.obs['cell_or_nuclei']='nuclei'\n",
    "atac.obs.rename(columns={'Donor':'donor'},inplace=True)\n",
    "atac.obs['donor_cellnuc']=atac.obs['donor'].astype('str')+'_'+atac.obs['cell_or_nuclei'].astype('str')\n",
    "\n",
    "# add modality to .var\n",
    "atac.var['modality']='Peaks'\n",
    "\n",
    "atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "identical-indianapolis",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "snATAC      48098\n",
       "Multiome    31110\n",
       "Name: modality, dtype: int64"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.obs=atac.obs[['cellatac_clusters','cellatac_code', 'sangerID', 'dataset', 'donor', 'Region', 'barcode', 'oribarcode','donor_cellnuc']]\n",
    "atac.var=atac.var[['modality','peak_width', 'exon', 'gene', 'promoter', 'annotation', 'gene_name','gene_id', 'tss_distance', 'ENCODE_blacklist',]]\n",
    "\n",
    "# set 'modality' in .obs as 'snATAC' and 'Multiome'\n",
    "atac.obs['modality']=atac.obs['dataset'].copy()\n",
    "atac.obs.replace({'modality':{\n",
    "    'Multiome_1':'Multiome',\n",
    "    'Multiome_2':'Multiome',\n",
    "    'LV':'snATAC',\n",
    "    'AX':'snATAC',\n",
    "    'SP':'snATAC',\n",
    "    'RA':'snATAC',\n",
    "    'LA':'snATAC',\n",
    "    'RV':'snATAC',\n",
    "}},inplace=True)\n",
    "atac.obs['modality'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "visible-publication",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array([1., 1., 1., 1., 1., 1., 1., 1., 1., 1.], dtype=float32)"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "atac.X.data[:10]"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "entertaining-quest",
   "metadata": {},
   "source": [
    "## Generate `adata_rna.h5ad`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "ecological-carbon",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 443004 × 31915\n",
       "    obs: 'sangerID', 'modality', 'donor', 'age_group', 'region', 'cell_or_nuclei', 'gender', 'type', 'cell_states', 'modality_fine', 'donor_cellnuc'\n",
       "    var: 'gene_ids-0', 'gene_ids-1', 'modality'"
      ]
     },
     "execution_count": 10,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset scnRNA data\n",
    "adata_rna=rna[rna.obs['modality']=='scnRNA']\n",
    "adata_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "sudden-summary",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 13822 × 31915\n",
       "    obs: 'sangerID', 'modality', 'donor', 'age_group', 'region', 'cell_or_nuclei', 'gender', 'type', 'cell_states', 'modality_fine', 'donor_cellnuc', 'batch'\n",
       "    var: 'modality'"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# downsize adata_rna per 'cell_states'\n",
    "adata_rna=downsize_cell(adata_rna,'cell_states',n_down=200, seed=123)\n",
    "adata_rna.var=adata_rna.var[['modality']].copy()\n",
    "adata_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "numeric-brass",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sangerID' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'donor' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'age_group' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'region' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'cell_or_nuclei' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'type' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'donor_cellnuc' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'modality' as categorical\n"
     ]
    }
   ],
   "source": [
    "# save\n",
    "adata_rna.write('/nfs/team205/kk18/data/6region_v2/MultiVI/adata_rna_downsized.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "blond-treaty",
   "metadata": {},
   "source": [
    "## Generate `adata_atac.h5ad`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "bright-islam",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "View of AnnData object with n_obs × n_vars = 48098 × 102627\n",
       "    obs: 'cellatac_clusters', 'cellatac_code', 'sangerID', 'dataset', 'donor', 'Region', 'barcode', 'oribarcode', 'donor_cellnuc', 'modality'\n",
       "    var: 'modality', 'peak_width', 'exon', 'gene', 'promoter', 'annotation', 'gene_name', 'gene_id', 'tss_distance', 'ENCODE_blacklist'\n",
       "    layers: 'binary_raw'"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_atac=atac[atac.obs['modality']=='snATAC']\n",
    "adata_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "several-birmingham",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 11170 × 102627\n",
       "    obs: 'cellatac_clusters', 'cellatac_code', 'sangerID', 'dataset', 'donor', 'Region', 'barcode', 'oribarcode', 'donor_cellnuc', 'modality', 'batch'\n",
       "    var: 'modality'\n",
       "    layers: 'binary_raw'"
      ]
     },
     "execution_count": 14,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# downsize adata_atac per 'cellatac_clusters'\n",
    "adata_atac.obs['cellatac_clusters']=adata_atac.obs['cellatac_clusters'].astype('category')\n",
    "adata_atac=downsize_cell(adata_atac,'cellatac_clusters',n_down=500, seed=123)\n",
    "adata_atac.var=adata_atac.var[['modality']].copy()\n",
    "adata_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "welcome-pantyhose",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'cellatac_code' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'sangerID' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'dataset' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'donor' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Region' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'barcode' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'donor_cellnuc' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'modality' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'modality' as categorical\n"
     ]
    }
   ],
   "source": [
    "adata_atac.write('/nfs/team205/kk18/data/6region_v2/MultiVI/adata_atac_downsized.h5ad')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "bacterial-surprise",
   "metadata": {},
   "source": [
    "## Generate `adata_paired.h5ad`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "japanese-lucas",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>SangerID</th>\n",
       "      <th>SampleID</th>\n",
       "      <th>Protocol</th>\n",
       "      <th>Target</th>\n",
       "      <th>Donor</th>\n",
       "      <th>Region</th>\n",
       "      <th>Gender</th>\n",
       "      <th>Age</th>\n",
       "      <th>Type</th>\n",
       "      <th>cellranger_outs</th>\n",
       "      <th>Combined_ID</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>A24-HEA-1-FTM-3</td>\n",
       "      <td>ATAC</td>\n",
       "      <td>10000</td>\n",
       "      <td>3</td>\n",
       "      <td>LV</td>\n",
       "      <td>Male</td>\n",
       "      <td>55-60</td>\n",
       "      <td>DBD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>HCAHeart9508820</td>\n",
       "      <td>A35-HEA-4-FTM-2</td>\n",
       "      <td>ATAC</td>\n",
       "      <td>10000</td>\n",
       "      <td>7</td>\n",
       "      <td>RA</td>\n",
       "      <td>Male</td>\n",
       "      <td>60-65</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9508628_HCAHeart9508820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>HCAHeart9508821</td>\n",
       "      <td>A35-HEA-1-FTM-2</td>\n",
       "      <td>ATAC</td>\n",
       "      <td>10000</td>\n",
       "      <td>7</td>\n",
       "      <td>LV</td>\n",
       "      <td>Male</td>\n",
       "      <td>60-65</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9508629_HCAHeart9508821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>A24-HEA-1-FTM-3</td>\n",
       "      <td>RNA</td>\n",
       "      <td>10000</td>\n",
       "      <td>3</td>\n",
       "      <td>LV</td>\n",
       "      <td>Male</td>\n",
       "      <td>55-60</td>\n",
       "      <td>DBD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>HCAHeart9508628</td>\n",
       "      <td>A35-HEA-4-FTM-2</td>\n",
       "      <td>RNA</td>\n",
       "      <td>10000</td>\n",
       "      <td>7</td>\n",
       "      <td>RA</td>\n",
       "      <td>Male</td>\n",
       "      <td>60-65</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9508628_HCAHeart9508820</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>HCAHeart9508629</td>\n",
       "      <td>A35-HEA-1-FTM-2</td>\n",
       "      <td>RNA</td>\n",
       "      <td>10000</td>\n",
       "      <td>7</td>\n",
       "      <td>LV</td>\n",
       "      <td>Male</td>\n",
       "      <td>60-65</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9508629_HCAHeart9508821</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>HCAHeart9845431</td>\n",
       "      <td>A39-HEA-1-FTM-4</td>\n",
       "      <td>RNA</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>LV</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845431_HCAHeart9917173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>HCAHeart9845432</td>\n",
       "      <td>A39-HEA-6-FTM-4</td>\n",
       "      <td>RNA</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>AX</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845432_HCAHeart9917174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>HCAHeart9845433</td>\n",
       "      <td>A39-HEA-3-FTM-4</td>\n",
       "      <td>RNA</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>LA</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845433_HCAHeart9917175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>HCAHeart9845434</td>\n",
       "      <td>A39-HEA-2-FTM-2</td>\n",
       "      <td>RNA</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>RV</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845434_HCAHeart9917176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>HCAHeart9845435</td>\n",
       "      <td>A39-HEA-5-FTM-4</td>\n",
       "      <td>RNA</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>SP</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845435_HCAHeart9917177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>HCAHeart9845436</td>\n",
       "      <td>A39-HEA-4-FTM-3</td>\n",
       "      <td>RNA</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>RA</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845436_HCAHeart9917178</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>HCAHeart9917173</td>\n",
       "      <td>A39-HEA-1-FTM-4</td>\n",
       "      <td>ATAC</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>LV</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845431_HCAHeart9917173</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>HCAHeart9917174</td>\n",
       "      <td>A39-HEA-6-FTM-4</td>\n",
       "      <td>ATAC</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>AX</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845432_HCAHeart9917174</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>HCAHeart9917175</td>\n",
       "      <td>A39-HEA-3-FTM-4</td>\n",
       "      <td>ATAC</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>LA</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845433_HCAHeart9917175</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>HCAHeart9917176</td>\n",
       "      <td>A39-HEA-2-FTM-2</td>\n",
       "      <td>ATAC</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>RV</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845434_HCAHeart9917176</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>HCAHeart9917177</td>\n",
       "      <td>A39-HEA-5-FTM-4</td>\n",
       "      <td>ATAC</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>SP</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845435_HCAHeart9917177</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>HCAHeart9917178</td>\n",
       "      <td>A39-HEA-4-FTM-3</td>\n",
       "      <td>ATAC</td>\n",
       "      <td>10000</td>\n",
       "      <td>8</td>\n",
       "      <td>RA</td>\n",
       "      <td>Male</td>\n",
       "      <td>45-50</td>\n",
       "      <td>DCD</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc101...</td>\n",
       "      <td>HCAHeart9845436_HCAHeart9917178</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "           SangerID         SampleID Protocol  Target  Donor Region Gender  \\\n",
       "0   HCAHeart9508819  A24-HEA-1-FTM-3     ATAC   10000      3     LV   Male   \n",
       "1   HCAHeart9508820  A35-HEA-4-FTM-2     ATAC   10000      7     RA   Male   \n",
       "2   HCAHeart9508821  A35-HEA-1-FTM-2     ATAC   10000      7     LV   Male   \n",
       "3   HCAHeart9508627  A24-HEA-1-FTM-3      RNA   10000      3     LV   Male   \n",
       "4   HCAHeart9508628  A35-HEA-4-FTM-2      RNA   10000      7     RA   Male   \n",
       "5   HCAHeart9508629  A35-HEA-1-FTM-2      RNA   10000      7     LV   Male   \n",
       "6   HCAHeart9845431  A39-HEA-1-FTM-4      RNA   10000      8     LV   Male   \n",
       "7   HCAHeart9845432  A39-HEA-6-FTM-4      RNA   10000      8     AX   Male   \n",
       "8   HCAHeart9845433  A39-HEA-3-FTM-4      RNA   10000      8     LA   Male   \n",
       "9   HCAHeart9845434  A39-HEA-2-FTM-2      RNA   10000      8     RV   Male   \n",
       "10  HCAHeart9845435  A39-HEA-5-FTM-4      RNA   10000      8     SP   Male   \n",
       "11  HCAHeart9845436  A39-HEA-4-FTM-3      RNA   10000      8     RA   Male   \n",
       "12  HCAHeart9917173  A39-HEA-1-FTM-4     ATAC   10000      8     LV   Male   \n",
       "13  HCAHeart9917174  A39-HEA-6-FTM-4     ATAC   10000      8     AX   Male   \n",
       "14  HCAHeart9917175  A39-HEA-3-FTM-4     ATAC   10000      8     LA   Male   \n",
       "15  HCAHeart9917176  A39-HEA-2-FTM-2     ATAC   10000      8     RV   Male   \n",
       "16  HCAHeart9917177  A39-HEA-5-FTM-4     ATAC   10000      8     SP   Male   \n",
       "17  HCAHeart9917178  A39-HEA-4-FTM-3     ATAC   10000      8     RA   Male   \n",
       "\n",
       "      Age Type                                    cellranger_outs  \\\n",
       "0   55-60  DBD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "1   60-65  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "2   60-65  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "3   55-60  DBD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "4   60-65  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "5   60-65  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "6   45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "7   45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "8   45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "9   45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "10  45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "11  45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "12  45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "13  45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "14  45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "15  45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "16  45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "17  45-50  DCD  /seq/illumina/cellranger-arc/cellranger-arc101...   \n",
       "\n",
       "                        Combined_ID  \n",
       "0   HCAHeart9508627_HCAHeart9508819  \n",
       "1   HCAHeart9508628_HCAHeart9508820  \n",
       "2   HCAHeart9508629_HCAHeart9508821  \n",
       "3   HCAHeart9508627_HCAHeart9508819  \n",
       "4   HCAHeart9508628_HCAHeart9508820  \n",
       "5   HCAHeart9508629_HCAHeart9508821  \n",
       "6   HCAHeart9845431_HCAHeart9917173  \n",
       "7   HCAHeart9845432_HCAHeart9917174  \n",
       "8   HCAHeart9845433_HCAHeart9917175  \n",
       "9   HCAHeart9845434_HCAHeart9917176  \n",
       "10  HCAHeart9845435_HCAHeart9917177  \n",
       "11  HCAHeart9845436_HCAHeart9917178  \n",
       "12  HCAHeart9845431_HCAHeart9917173  \n",
       "13  HCAHeart9845432_HCAHeart9917174  \n",
       "14  HCAHeart9845433_HCAHeart9917175  \n",
       "15  HCAHeart9845434_HCAHeart9917176  \n",
       "16  HCAHeart9845435_HCAHeart9917177  \n",
       "17  HCAHeart9845436_HCAHeart9917178  "
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# multiome metadata\n",
    "meta=pd.read_csv('/nfs/team205/kk18/data/cellatac/tic-1050/6reg-v2_Multiome_metadata.csv',index_col=0)\n",
    "meta"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "pretty-mediterranean",
   "metadata": {},
   "source": [
    "**Multiome, RNA**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "historic-badge",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "Trying to set attribute `.obs` of view, copying.\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:794: UserWarning: \n",
      "AnnData expects .obs.index to contain strings, but got values like:\n",
      "    [0, 1, 2, 3, 4]\n",
      "\n",
      "    Inferred to be: integer\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 30639 × 31915\n",
       "    obs: 'rna_sangerID', 'modality', 'donor', 'age_group', 'region', 'cell_or_nuclei', 'gender', 'type', 'cell_states', 'modality_fine', 'donor_cellnuc', 'barcode', 'Combined_ID'\n",
       "    var: 'gene_ids-0', 'gene_ids-1', 'modality'"
      ]
     },
     "execution_count": 17,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset Multiome-RNA\n",
    "multiome_rna=rna[rna.obs['modality']=='Multiome']\n",
    "\n",
    "# prepare GEM barcodes\n",
    "multiome_rna.obs['barcode']=multiome_rna.obs.index.copy()\n",
    "multiome_rna.obs['barcode']=[x.split('_')[1] for x in multiome_rna.obs.index]\n",
    "\n",
    "# add Combined_ID\n",
    "multiome_rna.obs.rename(columns={'sangerID':'rna_sangerID'},inplace=True)\n",
    "multiome_rna.obs=multiome_rna.obs.merge(meta[['SangerID','Combined_ID']],how='left',left_on='rna_sangerID',right_on='SangerID')\n",
    "multiome_rna.obs.drop(['SangerID'],axis=1,inplace=True)\n",
    "\n",
    "# set index with fullbarcode: Combined_ID + barcodes\n",
    "multiome_rna.obs['Combined_barcode']=multiome_rna.obs['Combined_ID'].astype('str')+ \\\n",
    "                                     '_'+ multiome_rna.obs['barcode'].astype('str')\n",
    "multiome_rna.obs.set_index('Combined_barcode',inplace=True)\n",
    "\n",
    "multiome_rna"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "veterinary-energy",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>rna_sangerID</th>\n",
       "      <th>modality</th>\n",
       "      <th>donor</th>\n",
       "      <th>age_group</th>\n",
       "      <th>region</th>\n",
       "      <th>cell_or_nuclei</th>\n",
       "      <th>gender</th>\n",
       "      <th>type</th>\n",
       "      <th>cell_states</th>\n",
       "      <th>modality_fine</th>\n",
       "      <th>donor_cellnuc</th>\n",
       "      <th>barcode</th>\n",
       "      <th>Combined_ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Combined_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3</td>\n",
       "      <td>55-60</td>\n",
       "      <td>LV</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>Male</td>\n",
       "      <td>DBD</td>\n",
       "      <td>PC2_atria</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>AAACATGCAGCAGGTA-1</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3</td>\n",
       "      <td>55-60</td>\n",
       "      <td>LV</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>Male</td>\n",
       "      <td>DBD</td>\n",
       "      <td>vCM3</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>AAACATGCATAGCTGC-1</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3</td>\n",
       "      <td>55-60</td>\n",
       "      <td>LV</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>Male</td>\n",
       "      <td>DBD</td>\n",
       "      <td>DOCK4+aMØ</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>AAACCAACAAGGTGCA-1</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3</td>\n",
       "      <td>55-60</td>\n",
       "      <td>LV</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>Male</td>\n",
       "      <td>DBD</td>\n",
       "      <td>vCM3</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>AAACCAACACTTAACG-1</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1</th>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3</td>\n",
       "      <td>55-60</td>\n",
       "      <td>LV</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>Male</td>\n",
       "      <td>DBD</td>\n",
       "      <td>vCM3</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>AAACCAACAGCCGCTA-1</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                       rna_sangerID  modality  \\\n",
       "Combined_barcode                                                                \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1  HCAHeart9508627  Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1  HCAHeart9508627  Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1  HCAHeart9508627  Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1  HCAHeart9508627  Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1  HCAHeart9508627  Multiome   \n",
       "\n",
       "                                                   donor age_group region  \\\n",
       "Combined_barcode                                                            \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1    D3     55-60     LV   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1    D3     55-60     LV   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1    D3     55-60     LV   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1    D3     55-60     LV   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1    D3     55-60     LV   \n",
       "\n",
       "                                                   cell_or_nuclei gender type  \\\n",
       "Combined_barcode                                                                \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1         nuclei   Male  DBD   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1         nuclei   Male  DBD   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1         nuclei   Male  DBD   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1         nuclei   Male  DBD   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1         nuclei   Male  DBD   \n",
       "\n",
       "                                                   cell_states modality_fine  \\\n",
       "Combined_barcode                                                               \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1   PC2_atria      Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1        vCM3      Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1   DOCK4+aMØ      Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1        vCM3      Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1        vCM3      Multiome   \n",
       "\n",
       "                                                   donor_cellnuc  \\\n",
       "Combined_barcode                                                   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1     D3_nuclei   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1     D3_nuclei   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1     D3_nuclei   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1     D3_nuclei   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1     D3_nuclei   \n",
       "\n",
       "                                                               barcode  \\\n",
       "Combined_barcode                                                         \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1  AAACATGCAGCAGGTA-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1  AAACATGCATAGCTGC-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1  AAACCAACAAGGTGCA-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1  AAACCAACACTTAACG-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1  AAACCAACAGCCGCTA-1   \n",
       "\n",
       "                                                                        Combined_ID  \n",
       "Combined_barcode                                                                     \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1  HCAHeart9508627_HCAHeart9508819  \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1  HCAHeart9508627_HCAHeart9508819  \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1  HCAHeart9508627_HCAHeart9508819  \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1  HCAHeart9508627_HCAHeart9508819  \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1  HCAHeart9508627_HCAHeart9508819  "
      ]
     },
     "execution_count": 18,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiome_rna.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "statistical-hawaiian",
   "metadata": {},
   "source": [
    "**Multiome, ATAC**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "academic-rapid",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:794: UserWarning: \n",
      "AnnData expects .obs.index to contain strings, but got values like:\n",
      "    [0, 1, 2, 3, 4]\n",
      "\n",
      "    Inferred to be: integer\n",
      "\n",
      "  value_idx = self._prep_dim_index(value.index, attr)\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 31110 × 102627\n",
       "    obs: 'cellatac_clusters', 'cellatac_code', 'atac_sangerID', 'dataset', 'donor', 'Region', 'barcode', 'oribarcode', 'donor_cellnuc', 'modality', 'Combined_ID'\n",
       "    var: 'modality', 'peak_width', 'exon', 'gene', 'promoter', 'annotation', 'gene_name', 'gene_id', 'tss_distance', 'ENCODE_blacklist'\n",
       "    layers: 'binary_raw'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# subset Multiome-ATAC\n",
    "multiome_atac=atac[atac.obs['modality']=='Multiome']\n",
    "\n",
    "# add Combined_ID\n",
    "multiome_atac.obs.rename(columns={'sangerID':'atac_sangerID'},inplace=True)\n",
    "multiome_atac.obs=multiome_atac.obs.merge(meta[['SangerID','Combined_ID']],how='left',left_on='atac_sangerID',right_on='SangerID')\n",
    "multiome_atac.obs.drop(['SangerID'],axis=1,inplace=True)\n",
    "\n",
    "# set index with fullbarcode: Combined_ID + barcodes\n",
    "multiome_atac.obs['Combined_barcode']=multiome_atac.obs['Combined_ID'].astype('str')+ \\\n",
    "                                     '_'+ multiome_atac.obs['barcode'].astype('str')\n",
    "multiome_atac.obs.set_index('Combined_barcode',inplace=True)\n",
    "\n",
    "multiome_atac"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "comfortable-holocaust",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cellatac_clusters</th>\n",
       "      <th>cellatac_code</th>\n",
       "      <th>atac_sangerID</th>\n",
       "      <th>dataset</th>\n",
       "      <th>donor</th>\n",
       "      <th>Region</th>\n",
       "      <th>barcode</th>\n",
       "      <th>oribarcode</th>\n",
       "      <th>donor_cellnuc</th>\n",
       "      <th>modality</th>\n",
       "      <th>Combined_ID</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Combined_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1</th>\n",
       "      <td>21</td>\n",
       "      <td>22</td>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>Multiome_1</td>\n",
       "      <td>D3</td>\n",
       "      <td>LV</td>\n",
       "      <td>AAACATGCAGCAGGTA-1</td>\n",
       "      <td>22-AAACATGCAGCAGGTA-1</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1</th>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>Multiome_1</td>\n",
       "      <td>D3</td>\n",
       "      <td>LV</td>\n",
       "      <td>AAACATGCATAGCTGC-1</td>\n",
       "      <td>22-AAACATGCATAGCTGC-1</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1</th>\n",
       "      <td>6</td>\n",
       "      <td>22</td>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>Multiome_1</td>\n",
       "      <td>D3</td>\n",
       "      <td>LV</td>\n",
       "      <td>AAACCAACAAGGTGCA-1</td>\n",
       "      <td>22-AAACCAACAAGGTGCA-1</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1</th>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>Multiome_1</td>\n",
       "      <td>D3</td>\n",
       "      <td>LV</td>\n",
       "      <td>AAACCAACACTTAACG-1</td>\n",
       "      <td>22-AAACCAACACTTAACG-1</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1</th>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>Multiome_1</td>\n",
       "      <td>D3</td>\n",
       "      <td>LV</td>\n",
       "      <td>AAACCAACAGCCGCTA-1</td>\n",
       "      <td>22-AAACCAACAGCCGCTA-1</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    cellatac_clusters  \\\n",
       "Combined_barcode                                                        \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1                 21   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1                 13   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1                  6   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1                 13   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1                 13   \n",
       "\n",
       "                                                   cellatac_code  \\\n",
       "Combined_barcode                                                   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1            22   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1            22   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1            22   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1            22   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1            22   \n",
       "\n",
       "                                                      atac_sangerID  \\\n",
       "Combined_barcode                                                      \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1  HCAHeart9508819   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1  HCAHeart9508819   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1  HCAHeart9508819   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1  HCAHeart9508819   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1  HCAHeart9508819   \n",
       "\n",
       "                                                       dataset donor Region  \\\n",
       "Combined_barcode                                                              \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1  Multiome_1    D3     LV   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1  Multiome_1    D3     LV   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1  Multiome_1    D3     LV   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1  Multiome_1    D3     LV   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1  Multiome_1    D3     LV   \n",
       "\n",
       "                                                               barcode  \\\n",
       "Combined_barcode                                                         \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1  AAACATGCAGCAGGTA-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1  AAACATGCATAGCTGC-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1  AAACCAACAAGGTGCA-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1  AAACCAACACTTAACG-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1  AAACCAACAGCCGCTA-1   \n",
       "\n",
       "                                                               oribarcode  \\\n",
       "Combined_barcode                                                            \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1  22-AAACATGCAGCAGGTA-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1  22-AAACATGCATAGCTGC-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1  22-AAACCAACAAGGTGCA-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1  22-AAACCAACACTTAACG-1   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1  22-AAACCAACAGCCGCTA-1   \n",
       "\n",
       "                                                   donor_cellnuc  modality  \\\n",
       "Combined_barcode                                                             \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1     D3_nuclei  Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1     D3_nuclei  Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1     D3_nuclei  Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1     D3_nuclei  Multiome   \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1     D3_nuclei  Multiome   \n",
       "\n",
       "                                                                        Combined_ID  \n",
       "Combined_barcode                                                                     \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCAGCAGGTA-1  HCAHeart9508627_HCAHeart9508819  \n",
       "HCAHeart9508627_HCAHeart9508819_AAACATGCATAGCTGC-1  HCAHeart9508627_HCAHeart9508819  \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAAGGTGCA-1  HCAHeart9508627_HCAHeart9508819  \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACACTTAACG-1  HCAHeart9508627_HCAHeart9508819  \n",
       "HCAHeart9508627_HCAHeart9508819_AAACCAACAGCCGCTA-1  HCAHeart9508627_HCAHeart9508819  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiome_atac.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "foreign-sound",
   "metadata": {},
   "source": [
    "**Concatenate Multiome-RNA and Multiome-ATAC**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "amino-college",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "30639"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# take intersecting barcodes between \n",
    "barcodes_inter=list(set(multiome_rna.obs_names).intersection(multiome_atac.obs_names))\n",
    "len(barcodes_inter)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "acute-tennessee",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "True"
      ]
     },
     "execution_count": 22,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "multiome_rna=multiome_rna[barcodes_inter,:]\n",
    "multiome_atac=multiome_atac[barcodes_inter,:]\n",
    "all(multiome_rna.obs_names==multiome_atac.obs_names)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "outdoor-aberdeen",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 30639 × 134542\n",
       "    obs: 'Combined_ID', 'rna_sangerID', 'barcode', 'donor', 'age_group', 'region', 'cell_or_nuclei', 'gender', 'type', 'cell_states', 'modality', 'donor_cellnuc', 'atac_sangerID', 'cellatac_clusters', 'cellatac_code'\n",
       "    var: 'modality', 'gene_ids-0', 'gene_ids-1', 'peak_width', 'exon', 'gene', 'promoter', 'annotation', 'gene_name', 'gene_id', 'tss_distance', 'ENCODE_blacklist'"
      ]
     },
     "execution_count": 23,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_paired=anndata.AnnData(\n",
    "    X=scipy.sparse.hstack([multiome_rna.X,multiome_atac.X]).tocsr(), # concatenate sparse matrix and convert to Compressed Sparse Row format\n",
    "    obs=pd.concat([multiome_rna.obs[['Combined_ID','rna_sangerID','barcode','donor', \n",
    "                                     'age_group','region','cell_or_nuclei', 'gender', 'type', 'cell_states','modality','donor_cellnuc']], \\\n",
    "                   multiome_atac.obs[['atac_sangerID','cellatac_clusters','cellatac_code',]]\n",
    "                  ],axis=1),\n",
    "    var=pd.concat([multiome_rna.var[['modality','gene_ids-0', 'gene_ids-1']], \\\n",
    "                   multiome_atac.var[['modality','peak_width', 'exon', 'gene', 'promoter', 'annotation', 'gene_name','gene_id', 'tss_distance', 'ENCODE_blacklist']]\n",
    "                  ],axis=0)\n",
    ")\n",
    "adata_paired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "fluid-former",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Combined_ID</th>\n",
       "      <th>rna_sangerID</th>\n",
       "      <th>barcode</th>\n",
       "      <th>donor</th>\n",
       "      <th>age_group</th>\n",
       "      <th>region</th>\n",
       "      <th>cell_or_nuclei</th>\n",
       "      <th>gender</th>\n",
       "      <th>type</th>\n",
       "      <th>cell_states</th>\n",
       "      <th>modality</th>\n",
       "      <th>donor_cellnuc</th>\n",
       "      <th>atac_sangerID</th>\n",
       "      <th>cellatac_clusters</th>\n",
       "      <th>cellatac_code</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Combined_barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1</th>\n",
       "      <td>HCAHeart9508627_HCAHeart9508819</td>\n",
       "      <td>HCAHeart9508627</td>\n",
       "      <td>CCGTTGCGTTAGACCA-1</td>\n",
       "      <td>D3</td>\n",
       "      <td>55-60</td>\n",
       "      <td>LV</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>Male</td>\n",
       "      <td>DBD</td>\n",
       "      <td>EC10_CMC-like</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D3_nuclei</td>\n",
       "      <td>HCAHeart9508819</td>\n",
       "      <td>13</td>\n",
       "      <td>22</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1</th>\n",
       "      <td>HCAHeart9845433_HCAHeart9917175</td>\n",
       "      <td>HCAHeart9845433</td>\n",
       "      <td>CTAGGACGTCACAGAC-1</td>\n",
       "      <td>D8</td>\n",
       "      <td>45-50</td>\n",
       "      <td>LA</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>Male</td>\n",
       "      <td>DCD</td>\n",
       "      <td>aCM1</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D8_nuclei</td>\n",
       "      <td>HCAHeart9917175</td>\n",
       "      <td>12</td>\n",
       "      <td>27</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1</th>\n",
       "      <td>HCAHeart9508628_HCAHeart9508820</td>\n",
       "      <td>HCAHeart9508628</td>\n",
       "      <td>ACAACAACATCCTGAA-1</td>\n",
       "      <td>D7</td>\n",
       "      <td>60-65</td>\n",
       "      <td>RA</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>Male</td>\n",
       "      <td>DCD</td>\n",
       "      <td>FB3</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D7_nuclei</td>\n",
       "      <td>HCAHeart9508820</td>\n",
       "      <td>10</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1</th>\n",
       "      <td>HCAHeart9508628_HCAHeart9508820</td>\n",
       "      <td>HCAHeart9508628</td>\n",
       "      <td>CCTCAATGTTAGTACG-1</td>\n",
       "      <td>D7</td>\n",
       "      <td>60-65</td>\n",
       "      <td>RA</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>Male</td>\n",
       "      <td>DCD</td>\n",
       "      <td>aCM2</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D7_nuclei</td>\n",
       "      <td>HCAHeart9508820</td>\n",
       "      <td>4</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1</th>\n",
       "      <td>HCAHeart9508628_HCAHeart9508820</td>\n",
       "      <td>HCAHeart9508628</td>\n",
       "      <td>CAATGACTCGGTTTGG-1</td>\n",
       "      <td>D7</td>\n",
       "      <td>60-65</td>\n",
       "      <td>RA</td>\n",
       "      <td>nuclei</td>\n",
       "      <td>Male</td>\n",
       "      <td>DCD</td>\n",
       "      <td>aCM4</td>\n",
       "      <td>Multiome</td>\n",
       "      <td>D7_nuclei</td>\n",
       "      <td>HCAHeart9508820</td>\n",
       "      <td>11</td>\n",
       "      <td>23</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                                        Combined_ID  \\\n",
       "Combined_barcode                                                                      \n",
       "HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1  HCAHeart9508627_HCAHeart9508819   \n",
       "HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1  HCAHeart9845433_HCAHeart9917175   \n",
       "HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1  HCAHeart9508628_HCAHeart9508820   \n",
       "HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1  HCAHeart9508628_HCAHeart9508820   \n",
       "HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1  HCAHeart9508628_HCAHeart9508820   \n",
       "\n",
       "                                                       rna_sangerID  \\\n",
       "Combined_barcode                                                      \n",
       "HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1  HCAHeart9508627   \n",
       "HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1  HCAHeart9845433   \n",
       "HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1  HCAHeart9508628   \n",
       "HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1  HCAHeart9508628   \n",
       "HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1  HCAHeart9508628   \n",
       "\n",
       "                                                               barcode donor  \\\n",
       "Combined_barcode                                                               \n",
       "HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1  CCGTTGCGTTAGACCA-1    D3   \n",
       "HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1  CTAGGACGTCACAGAC-1    D8   \n",
       "HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1  ACAACAACATCCTGAA-1    D7   \n",
       "HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1  CCTCAATGTTAGTACG-1    D7   \n",
       "HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1  CAATGACTCGGTTTGG-1    D7   \n",
       "\n",
       "                                                   age_group region  \\\n",
       "Combined_barcode                                                      \n",
       "HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1     55-60     LV   \n",
       "HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1     45-50     LA   \n",
       "HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1     60-65     RA   \n",
       "HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1     60-65     RA   \n",
       "HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1     60-65     RA   \n",
       "\n",
       "                                                   cell_or_nuclei gender type  \\\n",
       "Combined_barcode                                                                \n",
       "HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1         nuclei   Male  DBD   \n",
       "HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1         nuclei   Male  DCD   \n",
       "HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1         nuclei   Male  DCD   \n",
       "HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1         nuclei   Male  DCD   \n",
       "HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1         nuclei   Male  DCD   \n",
       "\n",
       "                                                      cell_states  modality  \\\n",
       "Combined_barcode                                                              \n",
       "HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1  EC10_CMC-like  Multiome   \n",
       "HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1           aCM1  Multiome   \n",
       "HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1            FB3  Multiome   \n",
       "HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1           aCM2  Multiome   \n",
       "HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1           aCM4  Multiome   \n",
       "\n",
       "                                                   donor_cellnuc  \\\n",
       "Combined_barcode                                                   \n",
       "HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1     D3_nuclei   \n",
       "HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1     D8_nuclei   \n",
       "HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1     D7_nuclei   \n",
       "HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1     D7_nuclei   \n",
       "HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1     D7_nuclei   \n",
       "\n",
       "                                                      atac_sangerID  \\\n",
       "Combined_barcode                                                      \n",
       "HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1  HCAHeart9508819   \n",
       "HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1  HCAHeart9917175   \n",
       "HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1  HCAHeart9508820   \n",
       "HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1  HCAHeart9508820   \n",
       "HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1  HCAHeart9508820   \n",
       "\n",
       "                                                    cellatac_clusters  \\\n",
       "Combined_barcode                                                        \n",
       "HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1                 13   \n",
       "HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1                 12   \n",
       "HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1                 10   \n",
       "HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1                  4   \n",
       "HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1                 11   \n",
       "\n",
       "                                                   cellatac_code  \n",
       "Combined_barcode                                                  \n",
       "HCAHeart9508627_HCAHeart9508819_CCGTTGCGTTAGACCA-1            22  \n",
       "HCAHeart9845433_HCAHeart9917175_CTAGGACGTCACAGAC-1            27  \n",
       "HCAHeart9508628_HCAHeart9508820_ACAACAACATCCTGAA-1            23  \n",
       "HCAHeart9508628_HCAHeart9508820_CCTCAATGTTAGTACG-1            23  \n",
       "HCAHeart9508628_HCAHeart9508820_CAATGACTCGGTTTGG-1            23  "
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_paired.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "suspected-omaha",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 7517 × 134542\n",
       "    obs: 'Combined_ID', 'rna_sangerID', 'barcode', 'donor', 'age_group', 'region', 'cell_or_nuclei', 'gender', 'type', 'cell_states', 'modality', 'donor_cellnuc', 'atac_sangerID', 'cellatac_clusters', 'cellatac_code', 'batch'\n",
       "    var: 'modality'"
      ]
     },
     "execution_count": 25,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# downsize adata_paired per 'cell_states'\n",
    "adata_paired=downsize_cell(adata_paired,'cell_states',n_down=200, seed=123)\n",
    "adata_paired.var=adata_paired.var[['modality']].copy()\n",
    "adata_paired"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "exotic-fiction",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'Combined_ID' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'rna_sangerID' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'barcode' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'donor' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'age_group' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'region' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'type' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'donor_cellnuc' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'atac_sangerID' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'cellatac_code' as categorical\n",
      "/home/jovyan/my-conda-envs/scvi/lib/python3.8/site-packages/anndata/_core/anndata.py:1220: FutureWarning: The `inplace` parameter in pandas.Categorical.reorder_categories is deprecated and will be removed in a future version. Removing unused categories will always return a new Categorical object.\n",
      "  c.reorder_categories(natsorted(c.categories), inplace=True)\n",
      "... storing 'modality' as categorical\n"
     ]
    }
   ],
   "source": [
    "# save\n",
    "adata_paired.write('/nfs/team205/kk18/data/6region_v2/MultiVI/adata_paired_downsized.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "desirable-undergraduate",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "designing-gentleman",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "scvi",
   "language": "python",
   "name": "scvi"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.10"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
