{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "located-mother",
   "metadata": {},
   "source": [
    "* only SAN/AVN cells\n",
    "* only New data"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "descending-holocaust",
   "metadata": {},
   "source": [
    "## Import modules"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "physical-taste",
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "import scanpy as sc\n",
    "import matplotlib.pyplot as plt\n",
    "import scrublet as scr\n",
    "import session_info"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "connected-leeds",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<details>\n",
       "<summary>Click to view session information</summary>\n",
       "<pre>\n",
       "-----\n",
       "matplotlib          3.3.4\n",
       "numpy               1.20.1\n",
       "pandas              1.1.4\n",
       "scanpy              1.7.1\n",
       "scrublet            NA\n",
       "session_info        1.0.0\n",
       "-----\n",
       "</pre>\n",
       "<details>\n",
       "<summary>Click to view modules imported as dependencies</summary>\n",
       "<pre>\n",
       "PIL                 8.1.2\n",
       "anndata             0.7.5\n",
       "anyio               NA\n",
       "attr                20.3.0\n",
       "babel               2.9.0\n",
       "backcall            0.2.0\n",
       "brotli              NA\n",
       "cairo               1.20.0\n",
       "certifi             2020.12.05\n",
       "cffi                1.14.5\n",
       "chardet             4.0.0\n",
       "cloudpickle         1.6.0\n",
       "colorama            0.4.4\n",
       "cycler              0.10.0\n",
       "cython_runtime      NA\n",
       "cytoolz             0.11.0\n",
       "dask                2021.03.1\n",
       "dateutil            2.8.1\n",
       "decorator           4.4.2\n",
       "fsspec              0.8.7\n",
       "get_version         2.1\n",
       "google              NA\n",
       "h5py                2.10.0\n",
       "idna                2.10\n",
       "igraph              0.8.3\n",
       "ipykernel           5.5.0\n",
       "ipython_genutils    0.2.0\n",
       "jedi                0.18.0\n",
       "jinja2              2.11.3\n",
       "joblib              1.0.1\n",
       "json5               NA\n",
       "jsonschema          3.2.0\n",
       "jupyter_server      1.4.1\n",
       "jupyterlab_server   2.3.0\n",
       "kiwisolver          1.3.1\n",
       "legacy_api_wrap     0.0.0\n",
       "leidenalg           0.8.3\n",
       "llvmlite            0.34.0\n",
       "louvain             0.7.0\n",
       "markupsafe          1.1.1\n",
       "mpl_toolkits        NA\n",
       "natsort             7.1.1\n",
       "nbclassic           NA\n",
       "nbformat            5.1.2\n",
       "numba               0.51.2\n",
       "numexpr             2.7.3\n",
       "packaging           20.9\n",
       "parso               0.8.1\n",
       "pexpect             4.8.0\n",
       "pickleshare         0.7.5\n",
       "pkg_resources       NA\n",
       "prometheus_client   NA\n",
       "prompt_toolkit      3.0.16\n",
       "psutil              5.8.0\n",
       "ptyprocess          0.7.0\n",
       "pvectorc            NA\n",
       "pyarrow             0.16.0\n",
       "pygments            2.8.0\n",
       "pyparsing           2.4.7\n",
       "pyrsistent          NA\n",
       "pytoml              NA\n",
       "pytz                2021.1\n",
       "requests            2.25.1\n",
       "ruamel              NA\n",
       "scipy               1.6.1\n",
       "send2trash          NA\n",
       "setuptools_scm      NA\n",
       "simplejson          3.17.2\n",
       "sinfo               0.3.1\n",
       "six                 1.15.0\n",
       "sklearn             0.22\n",
       "sniffio             1.2.0\n",
       "socks               1.7.1\n",
       "sphinxcontrib       NA\n",
       "storemagic          NA\n",
       "tables              3.6.1\n",
       "tblib               1.7.0\n",
       "texttable           1.6.3\n",
       "tlz                 0.11.0\n",
       "toolz               0.11.1\n",
       "tornado             6.1\n",
       "traitlets           5.0.5\n",
       "typing_extensions   NA\n",
       "uritemplate         3.0.1\n",
       "urllib3             1.26.3\n",
       "wcwidth             0.2.5\n",
       "yaml                5.1.2\n",
       "zmq                 22.0.3\n",
       "</pre>\n",
       "</details> <!-- seems like this ends pre, so might as well be explicit -->\n",
       "<pre>\n",
       "-----\n",
       "IPython             7.21.0\n",
       "jupyter_client      6.1.11\n",
       "jupyter_core        4.7.1\n",
       "jupyterlab          3.0.9\n",
       "notebook            6.2.0\n",
       "-----\n",
       "Python 3.8.8 | packaged by conda-forge | (default, Feb 20 2021, 16:22:27) [GCC 9.3.0]\n",
       "Linux-4.15.0-156-generic-x86_64-with-glibc2.10\n",
       "-----\n",
       "Session information updated at 2022-12-08 13:49\n",
       "</pre>\n",
       "</details>"
      ],
      "text/plain": [
       "<IPython.core.display.HTML object>"
      ]
     },
     "execution_count": 2,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "sc.settings.verbosity = 3\n",
    "sc.settings.set_figure_params(dpi = 160, color_map = 'RdPu', dpi_save = 180, vector_friendly = True, format = 'svg')\n",
    "session_info.show()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "twenty-immigration",
   "metadata": {},
   "source": [
    "## Functions"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "addressed-monkey",
   "metadata": {},
   "outputs": [],
   "source": [
    "\"\"\"Functions for downstream work with outputs of remove-background.\"\"\"\n",
    "\n",
    "# to read cellbender outputs<br>\n",
    "# https://github.com/broadinstitute/CellBender/issues/57\n",
    "\n",
    "import tables\n",
    "import numpy as np\n",
    "import scipy.sparse as sp\n",
    "from typing import Dict\n",
    "\n",
    "\n",
    "def dict_from_h5(file: str) -> Dict[str, np.ndarray]:\n",
    "    \"\"\"Read in everything from an h5 file and put into a dictionary.\"\"\"\n",
    "    d = {}\n",
    "    with tables.open_file(file) as f:\n",
    "        # read in everything\n",
    "        for array in f.walk_nodes(\"/\", \"Array\"):\n",
    "            d[array.name] = array.read()\n",
    "    return d\n",
    "\n",
    "\n",
    "def anndata_from_h5(file: str,\n",
    "                    analyzed_barcodes_only: bool = True) -> 'anndata.AnnData':\n",
    "    \"\"\"Load an output h5 file into an AnnData object for downstream work.\n",
    "\n",
    "    Args:\n",
    "        file: The h5 file\n",
    "        analyzed_barcodes_only: False to load all barcodes, so that the size of\n",
    "            the AnnData object will match the size of the input raw count matrix.\n",
    "            True to load a limited set of barcodes: only those analyzed by the\n",
    "            algorithm. This allows relevant latent variables to be loaded\n",
    "            properly into adata.obs and adata.obsm, rather than adata.uns.\n",
    "\n",
    "    Returns:\n",
    "        adata: The anndata object, populated with inferred latent variables\n",
    "            and metadata.\n",
    "\n",
    "    \"\"\"\n",
    "\n",
    "    try:\n",
    "        import anndata\n",
    "    except ImportError:\n",
    "        raise ImportError('The anndata package must be installed to use the '\n",
    "                          'function anndata_from_h5()')\n",
    "\n",
    "    d = dict_from_h5(file)\n",
    "    X = sp.csc_matrix((d.pop('data'), d.pop('indices'), d.pop('indptr')),\n",
    "                      shape=d.pop('shape')).transpose().tocsr()\n",
    "\n",
    "    if analyzed_barcodes_only:\n",
    "        if 'barcodes_analyzed_inds' in d.keys():\n",
    "            X = X[d['barcodes_analyzed_inds'], :]\n",
    "            d['barcodes'] = d['barcodes'][d['barcodes_analyzed_inds']]\n",
    "        elif 'barcode_indices_for_latents' in d.keys():\n",
    "            X = X[d['barcode_indices_for_latents'], :]\n",
    "            d['barcodes'] = d['barcodes'][d['barcode_indices_for_latents']]\n",
    "        else:\n",
    "            print('Warning: analyzed_barcodes_only=True, but the key '\n",
    "                  '\"barcodes_analyzed_inds\" or \"barcode_indices_for_latents\" '\n",
    "                  'is missing from the h5 file. '\n",
    "                  'Will output all barcodes, and proceed as if '\n",
    "                  'analyzed_barcodes_only=False')\n",
    "\n",
    "    # Construct the count matrix.\n",
    "    adata = anndata.AnnData(X=X,\n",
    "                            obs={'barcode': d.pop('barcodes').astype(str)},\n",
    "                            var={'gene_name': (d.pop('gene_names') if 'gene_names' in d.keys()\n",
    "                                               else d.pop('name')).astype(str)})\n",
    "    adata.obs.set_index('barcode', inplace=True)\n",
    "    adata.var.set_index('gene_name', inplace=True)\n",
    "\n",
    "    # Add other information to the adata object in the appropriate slot.\n",
    "    for key, value in d.items():\n",
    "        try:\n",
    "            value = np.asarray(value)\n",
    "            if len(value.shape) == 0:\n",
    "                adata.uns[key] = value\n",
    "            elif value.shape[0] == X.shape[0]:\n",
    "                if (len(value.shape) < 2) or (value.shape[1] < 2):\n",
    "                    adata.obs[key] = value\n",
    "                else:\n",
    "                    adata.obsm[key] = value\n",
    "            elif value.shape[0] == X.shape[1]:\n",
    "                if value.dtype.name.startswith('bytes'):\n",
    "                    adata.var[key] = value.astype(str)\n",
    "                else:\n",
    "                    adata.var[key] = value\n",
    "            else:\n",
    "                adata.uns[key] = value\n",
    "        except Exception:\n",
    "            print('Unable to load data into AnnData: ', key, value, type(value))\n",
    "\n",
    "    if analyzed_barcodes_only:\n",
    "        for col in adata.obs.columns[adata.obs.columns.str.startswith('barcodes_analyzed')\n",
    "                                     | adata.obs.columns.str.startswith('barcode_indices')]:\n",
    "            try:\n",
    "                del adata.obs[col]\n",
    "            except Exception:\n",
    "                pass\n",
    "\n",
    "    return adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "tender-august",
   "metadata": {},
   "outputs": [],
   "source": [
    "# from https://github.com/Teichlab/mapcloud/blob/master/scripts/starsolo/postprocess.py\n",
    "\n",
    "from statsmodels.stats.multitest import multipletests\n",
    "# from emptydrops.matrix import CountMatrix\n",
    "# from emptydrops import find_nonambient_barcodes\n",
    "import scrublet as scr\n",
    "import scanpy as sc\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "import scipy\n",
    "\n",
    "#some functions that Ni uses in scanpy scripts to run scrublet\n",
    "#which in turn are inspired by my original notebook on the matter\n",
    "#(extracted from scanpy_scripts 0.2.10 to get around scanpy version incompatibility)\n",
    "def test_outlier(x, upper_mad_only=True):\n",
    "\tmed = np.median(x)\n",
    "\tif upper_mad_only:\n",
    "\t\tmad = np.median(x[x>med] - med) * 1.4826\n",
    "\telse:\n",
    "\t\tmad = np.median(np.abs(x - med)) * 1.4826\n",
    "\tpvals = 1 - scipy.stats.norm.cdf(x, loc=med, scale=mad)\n",
    "\tbh_pvals = multipletests(pvals, method='fdr_bh')[1]\n",
    "\treturn pvals, bh_pvals\n",
    "\n",
    "def run_scrublet(adata, resolution_function=None):\n",
    "\told_verbosity = sc.settings.verbosity\n",
    "\tsc.settings.verbosity = 1\n",
    "\tif resolution_function is None:\n",
    "\t\tresolution_function = lambda x: np.maximum(np.maximum(np.log10(x)-1, 0)**2, 0.1)\n",
    "\tscrub = scr.Scrublet(adata.X)\n",
    "\t#this has the potential to brick for poor quality data\n",
    "\t#if so, abort it and everything downstream\n",
    "\ttry:\n",
    "\t\tds, pd = scrub.scrub_doublets(verbose=False)\n",
    "\texcept:\n",
    "\t\treturn\n",
    "\tadata.obs['scrublet_score'] = ds\n",
    "\n",
    "\tadata_copy = adata.copy()\n",
    "\tsc.pp.filter_genes(adata_copy, min_cells=3)\n",
    "\tsc.pp.normalize_total(adata_copy, target_sum=1e4)\n",
    "\tsc.pp.log1p(adata_copy)\n",
    "\tsc.pp.highly_variable_genes(adata_copy, min_mean=0.0125, max_mean=3, min_disp=0.5, subset=True)\n",
    "\tsc.pp.scale(adata_copy, zero_center=False)\n",
    "\tsc.pp.pca(adata_copy, svd_solver='arpack', zero_center=False)\n",
    "\tsc.pp.neighbors(adata_copy, n_pcs=30)\n",
    "\tsc.tl.umap(adata_copy)\n",
    "\tsc.tl.leiden(adata_copy, resolution=1)\n",
    "\tfor clst in np.unique(adata_copy.obs['leiden']):\n",
    "\t\tclst_size = sum(adata_copy.obs['leiden'] == clst)\n",
    "\t\tsc.tl.leiden(adata_copy, restrict_to=('leiden', [clst]), resolution=resolution_function(clst_size), key_added='leiden_R')\n",
    "\t\tadata_copy.obs['leiden'] = adata_copy.obs['leiden_R']\n",
    "\tclst_meds = []\n",
    "\tfor clst in np.unique(adata_copy.obs['leiden']):\n",
    "\t\tk = adata_copy.obs['leiden'] == clst\n",
    "\t\tclst_med = np.median(adata_copy.obs.loc[k, 'scrublet_score'])\n",
    "\t\tadata_copy.obs.loc[k, 'cluster_scrublet_score'] = clst_med\n",
    "\t\tclst_meds.append(clst_med)\n",
    "\tclst_meds = np.array(clst_meds)\n",
    "\tpvals, bh_pvals = test_outlier(clst_meds)\n",
    "\tfor i, clst in enumerate(np.unique(adata_copy.obs['leiden'])):\n",
    "\t\tk = adata_copy.obs['leiden'] == clst\n",
    "\t\tadata_copy.obs.loc[k, 'pval'] = pvals[i]\n",
    "\t\tadata_copy.obs.loc[k, 'bh_pval'] = bh_pvals[i]\n",
    "\tsc.settings.verbosity = old_verbosity\n",
    "\t#need to also export the clustering, for soupx purposes\n",
    "\tadata.obs['scrublet_leiden'] = adata_copy.obs['leiden']\n",
    "\tadata.obs['scrublet_score'] = adata_copy.obs['scrublet_score']\n",
    "\tadata.obs['cluster_scrublet_score'] = adata_copy.obs['cluster_scrublet_score']\n",
    "\tadata.obs['doublet_pval'] = adata_copy.obs['pval']\n",
    "\tadata.obs['doublet_bh_pval'] = adata_copy.obs['bh_pval']\n",
    "\tdel adata_copy"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "tamil-bread",
   "metadata": {},
   "source": [
    "## Create anndata object"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "bright-demonstration",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Multiome-ATAC    17\n",
      "Multiome-RNA     17\n",
      "Visium            8\n",
      "Name: modality, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "# read in metadata\n",
    "metadata = pd.read_csv('/nfs/team205/heart/anndata_objects/8regions/metadata/HeartTeamSamples_Mappeddata_20221208.csv', sep = ',', index_col = None)\n",
    "metadata = metadata[metadata['Publication']=='8regions_revision']\n",
    "\n",
    "print(metadata['modality'].value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "based-british",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>CombinedID_Multiome</th>\n",
       "      <th>Publication</th>\n",
       "      <th>Foetal_or_Adult</th>\n",
       "      <th>Provider</th>\n",
       "      <th>Chemistry</th>\n",
       "      <th>Mapping_ver</th>\n",
       "      <th>Reference_genome</th>\n",
       "      <th>Mapping_date</th>\n",
       "      <th>Mapping_iRODS</th>\n",
       "      <th>...</th>\n",
       "      <th>region</th>\n",
       "      <th>region_finest</th>\n",
       "      <th>age</th>\n",
       "      <th>gender</th>\n",
       "      <th>facility</th>\n",
       "      <th>cell_or_nuclei</th>\n",
       "      <th>modality</th>\n",
       "      <th>kit_10x</th>\n",
       "      <th>flushed</th>\n",
       "      <th>status</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>473</th>\n",
       "      <td>HCAHeartST13146201</td>\n",
       "      <td>HCAHeartST13146201_HCAHeartST13129134</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>20-25</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>474</th>\n",
       "      <td>HCAHeartST13146202</td>\n",
       "      <td>HCAHeartST13146202_HCAHeartST13129135</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>20-25</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>475</th>\n",
       "      <td>HCAHeartST13146203</td>\n",
       "      <td>HCAHeartST13146203_HCAHeartST13129136</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>20-25</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>476</th>\n",
       "      <td>HCAHeartST13146204</td>\n",
       "      <td>HCAHeartST13146204_HCAHeartST13129137</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>45-50</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>477</th>\n",
       "      <td>HCAHeartST13146205</td>\n",
       "      <td>HCAHeartST13146205_HCAHeartST13129138</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>45-50</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>478</th>\n",
       "      <td>HCAHeartST13146206</td>\n",
       "      <td>HCAHeartST13146206_HCAHeartST13129139</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>45-50</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>479</th>\n",
       "      <td>HCAHeartST13146207</td>\n",
       "      <td>HCAHeartST13146207_HCAHeartST13129140</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>60-65</td>\n",
       "      <td>Female</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>480</th>\n",
       "      <td>HCAHeartST13146208</td>\n",
       "      <td>HCAHeartST13146208_HCAHeartST13129141</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>60-65</td>\n",
       "      <td>Female</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>543</th>\n",
       "      <td>HCAHeartST13168796</td>\n",
       "      <td>HCAHeartST13168796_HCAHeartST13167740</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>SAN</td>\n",
       "      <td>45-50</td>\n",
       "      <td>Female</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>548</th>\n",
       "      <td>HCAHeartST13189995</td>\n",
       "      <td>HCAHeartST13189995_HCAHeartST13188800</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>SAN</td>\n",
       "      <td>70-75</td>\n",
       "      <td>Female</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>549</th>\n",
       "      <td>HCAHeartST13189996</td>\n",
       "      <td>HCAHeartST13189996_HCAHeartST13188801</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>AVN</td>\n",
       "      <td>AVN</td>\n",
       "      <td>70-75</td>\n",
       "      <td>Female</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>550</th>\n",
       "      <td>HCAHeartST13189997</td>\n",
       "      <td>HCAHeartST13189997_HCAHeartST13188802</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>AVN</td>\n",
       "      <td>AVN</td>\n",
       "      <td>70-75</td>\n",
       "      <td>Female</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>581</th>\n",
       "      <td>HCAHeartST13180615</td>\n",
       "      <td>HCAHeartST13180615_HCAHeartST13177112</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>SAN</td>\n",
       "      <td>40-45</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>582</th>\n",
       "      <td>HCAHeartST13180616</td>\n",
       "      <td>HCAHeartST13180616_HCAHeartST13177113</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>SAN</td>\n",
       "      <td>40-45</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>583</th>\n",
       "      <td>HCAHeartST13180617</td>\n",
       "      <td>HCAHeartST13180617_HCAHeartST13177114</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>SAN</td>\n",
       "      <td>SAN</td>\n",
       "      <td>40-45</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>584</th>\n",
       "      <td>HCAHeartST13180618</td>\n",
       "      <td>HCAHeartST13180618_HCAHeartST13177115</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>AVN</td>\n",
       "      <td>AVN</td>\n",
       "      <td>40-45</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>585</th>\n",
       "      <td>HCAHeartST13180619</td>\n",
       "      <td>HCAHeartST13180619_HCAHeartST13177116</td>\n",
       "      <td>8regions_revision</td>\n",
       "      <td>Adult</td>\n",
       "      <td>Sanger, Heart</td>\n",
       "      <td>Single Cell Multiome ATAC + Gene Expression v1</td>\n",
       "      <td>cellranger-arc-2.0.0</td>\n",
       "      <td>GRCh38-2020-A-2.0.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>/seq/illumina/cellranger-arc/cellranger-arc200...</td>\n",
       "      <td>...</td>\n",
       "      <td>AVN</td>\n",
       "      <td>AVN</td>\n",
       "      <td>40-45</td>\n",
       "      <td>Male</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>Healthy</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>17 rows × 33 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     ID                    CombinedID_Multiome  \\\n",
       "473  HCAHeartST13146201  HCAHeartST13146201_HCAHeartST13129134   \n",
       "474  HCAHeartST13146202  HCAHeartST13146202_HCAHeartST13129135   \n",
       "475  HCAHeartST13146203  HCAHeartST13146203_HCAHeartST13129136   \n",
       "476  HCAHeartST13146204  HCAHeartST13146204_HCAHeartST13129137   \n",
       "477  HCAHeartST13146205  HCAHeartST13146205_HCAHeartST13129138   \n",
       "478  HCAHeartST13146206  HCAHeartST13146206_HCAHeartST13129139   \n",
       "479  HCAHeartST13146207  HCAHeartST13146207_HCAHeartST13129140   \n",
       "480  HCAHeartST13146208  HCAHeartST13146208_HCAHeartST13129141   \n",
       "543  HCAHeartST13168796  HCAHeartST13168796_HCAHeartST13167740   \n",
       "548  HCAHeartST13189995  HCAHeartST13189995_HCAHeartST13188800   \n",
       "549  HCAHeartST13189996  HCAHeartST13189996_HCAHeartST13188801   \n",
       "550  HCAHeartST13189997  HCAHeartST13189997_HCAHeartST13188802   \n",
       "581  HCAHeartST13180615  HCAHeartST13180615_HCAHeartST13177112   \n",
       "582  HCAHeartST13180616  HCAHeartST13180616_HCAHeartST13177113   \n",
       "583  HCAHeartST13180617  HCAHeartST13180617_HCAHeartST13177114   \n",
       "584  HCAHeartST13180618  HCAHeartST13180618_HCAHeartST13177115   \n",
       "585  HCAHeartST13180619  HCAHeartST13180619_HCAHeartST13177116   \n",
       "\n",
       "           Publication Foetal_or_Adult       Provider  \\\n",
       "473  8regions_revision           Adult  Sanger, Heart   \n",
       "474  8regions_revision           Adult  Sanger, Heart   \n",
       "475  8regions_revision           Adult  Sanger, Heart   \n",
       "476  8regions_revision           Adult  Sanger, Heart   \n",
       "477  8regions_revision           Adult  Sanger, Heart   \n",
       "478  8regions_revision           Adult  Sanger, Heart   \n",
       "479  8regions_revision           Adult  Sanger, Heart   \n",
       "480  8regions_revision           Adult  Sanger, Heart   \n",
       "543  8regions_revision           Adult  Sanger, Heart   \n",
       "548  8regions_revision           Adult  Sanger, Heart   \n",
       "549  8regions_revision           Adult  Sanger, Heart   \n",
       "550  8regions_revision           Adult  Sanger, Heart   \n",
       "581  8regions_revision           Adult  Sanger, Heart   \n",
       "582  8regions_revision           Adult  Sanger, Heart   \n",
       "583  8regions_revision           Adult  Sanger, Heart   \n",
       "584  8regions_revision           Adult  Sanger, Heart   \n",
       "585  8regions_revision           Adult  Sanger, Heart   \n",
       "\n",
       "                                          Chemistry           Mapping_ver  \\\n",
       "473  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "474  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "475  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "476  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "477  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "478  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "479  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "480  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "543  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "548  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "549  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "550  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "581  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "582  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "583  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "584  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "585  Single Cell Multiome ATAC + Gene Expression v1  cellranger-arc-2.0.0   \n",
       "\n",
       "        Reference_genome Mapping_date  \\\n",
       "473  GRCh38-2020-A-2.0.0          NaN   \n",
       "474  GRCh38-2020-A-2.0.0          NaN   \n",
       "475  GRCh38-2020-A-2.0.0          NaN   \n",
       "476  GRCh38-2020-A-2.0.0          NaN   \n",
       "477  GRCh38-2020-A-2.0.0          NaN   \n",
       "478  GRCh38-2020-A-2.0.0          NaN   \n",
       "479  GRCh38-2020-A-2.0.0          NaN   \n",
       "480  GRCh38-2020-A-2.0.0          NaN   \n",
       "543  GRCh38-2020-A-2.0.0          NaN   \n",
       "548  GRCh38-2020-A-2.0.0          NaN   \n",
       "549  GRCh38-2020-A-2.0.0          NaN   \n",
       "550  GRCh38-2020-A-2.0.0          NaN   \n",
       "581  GRCh38-2020-A-2.0.0          NaN   \n",
       "582  GRCh38-2020-A-2.0.0          NaN   \n",
       "583  GRCh38-2020-A-2.0.0          NaN   \n",
       "584  GRCh38-2020-A-2.0.0          NaN   \n",
       "585  GRCh38-2020-A-2.0.0          NaN   \n",
       "\n",
       "                                         Mapping_iRODS  ... region  \\\n",
       "473  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "474  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "475  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "476  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "477  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "478  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "479  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "480  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "543  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "548  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "549  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    AVN   \n",
       "550  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    AVN   \n",
       "581  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "582  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "583  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    SAN   \n",
       "584  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    AVN   \n",
       "585  /seq/illumina/cellranger-arc/cellranger-arc200...  ...    AVN   \n",
       "\n",
       "    region_finest    age  gender facility cell_or_nuclei      modality  \\\n",
       "473            na  20-25    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "474            na  20-25    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "475            na  20-25    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "476            na  45-50    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "477            na  45-50    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "478            na  45-50    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "479            na  60-65  Female   Sanger         Nuclei  Multiome-RNA   \n",
       "480            na  60-65  Female   Sanger         Nuclei  Multiome-RNA   \n",
       "543           SAN  45-50  Female   Sanger         Nuclei  Multiome-RNA   \n",
       "548           SAN  70-75  Female   Sanger         Nuclei  Multiome-RNA   \n",
       "549           AVN  70-75  Female   Sanger         Nuclei  Multiome-RNA   \n",
       "550           AVN  70-75  Female   Sanger         Nuclei  Multiome-RNA   \n",
       "581           SAN  40-45    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "582           SAN  40-45    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "583           SAN  40-45    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "584           AVN  40-45    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "585           AVN  40-45    Male   Sanger         Nuclei  Multiome-RNA   \n",
       "\n",
       "         kit_10x flushed   status  \n",
       "473  Multiome-v1     yes  Healthy  \n",
       "474  Multiome-v1     yes  Healthy  \n",
       "475  Multiome-v1     yes  Healthy  \n",
       "476  Multiome-v1     yes  Healthy  \n",
       "477  Multiome-v1     yes  Healthy  \n",
       "478  Multiome-v1     yes  Healthy  \n",
       "479  Multiome-v1     yes  Healthy  \n",
       "480  Multiome-v1     yes  Healthy  \n",
       "543  Multiome-v1     yes  Healthy  \n",
       "548  Multiome-v1     yes  Healthy  \n",
       "549  Multiome-v1     yes  Healthy  \n",
       "550  Multiome-v1     yes  Healthy  \n",
       "581  Multiome-v1     yes  Healthy  \n",
       "582  Multiome-v1     yes  Healthy  \n",
       "583  Multiome-v1     yes  Healthy  \n",
       "584  Multiome-v1     yes  Healthy  \n",
       "585  Multiome-v1     yes  Healthy  \n",
       "\n",
       "[17 rows x 33 columns]"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# select modality\n",
    "metadata = metadata[metadata['modality']=='Multiome-RNA']\n",
    "\n",
    "# select SAN\n",
    "metadata = metadata[metadata['region'].isin(['SAN','AVN'])]\n",
    "metadata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "correct-integrity",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13146201_HCAHeartST13129134\n",
      "HCAHeartST13146202_HCAHeartST13129135\n",
      "HCAHeartST13146203_HCAHeartST13129136\n",
      "HCAHeartST13146204_HCAHeartST13129137\n",
      "HCAHeartST13146205_HCAHeartST13129138\n",
      "HCAHeartST13146206_HCAHeartST13129139\n",
      "HCAHeartST13146207_HCAHeartST13129140\n",
      "HCAHeartST13146208_HCAHeartST13129141\n",
      "HCAHeartST13168796_HCAHeartST13167740\n",
      "HCAHeartST13189995_HCAHeartST13188800\n",
      "HCAHeartST13189996_HCAHeartST13188801\n",
      "HCAHeartST13189997_HCAHeartST13188802\n",
      "HCAHeartST13180615_HCAHeartST13177112\n",
      "HCAHeartST13180616_HCAHeartST13177113\n",
      "HCAHeartST13180617_HCAHeartST13177114\n",
      "HCAHeartST13180618_HCAHeartST13177115\n",
      "HCAHeartST13180619_HCAHeartST13177116\n"
     ]
    },
    {
     "data": {
      "text/plain": [
       "(146652, 36601)"
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "########### editted: put id to var_names ###########\n",
    "\n",
    "# read in\n",
    "adatas=[]\n",
    "for i in range(len(metadata)):\n",
    "    print(metadata.iloc[i]['combinedID'])\n",
    "    \n",
    "    path=metadata.iloc[i]['CellBender_out']\n",
    "    adatas.append(anndata_from_h5(path + '/' + path.split('/')[-1] + '_cellbender_out_filtered.h5', analyzed_barcodes_only=False))\n",
    "    del path\n",
    "    \n",
    "    # replace var_names with gene_id \n",
    "    adatas[i].var.reset_index(inplace=True)\n",
    "    adatas[i].var.set_index('id',inplace=True)\n",
    "    \n",
    "    # modify barcodes\n",
    "    adatas[i].obs.index = metadata.iloc[i]['combinedID'] + '_' + adatas[i].obs.index \n",
    "    \n",
    "    # add metadata\n",
    "    for col in ['sangerID','combinedID', 'donor', 'donor_type', 'region', 'region_finest', 'age',\n",
    "                'gender', 'facility', 'cell_or_nuclei', 'modality', 'kit_10x','flushed']:\n",
    "        adatas[i].obs[col] = metadata.iloc[i][col]\n",
    "\n",
    "\n",
    "# concatenate\n",
    "adata = adatas[0].concatenate(adatas[1:], index_unique = None, batch_key=None)\n",
    "adata.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "seeing-debate",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "HCAHeartST13180617_HCAHeartST13177114    16625\n",
       "HCAHeartST13180615_HCAHeartST13177112    16497\n",
       "HCAHeartST13189996_HCAHeartST13188801    15965\n",
       "HCAHeartST13180616_HCAHeartST13177113    12300\n",
       "HCAHeartST13180618_HCAHeartST13177115    10480\n",
       "HCAHeartST13189997_HCAHeartST13188802    10159\n",
       "HCAHeartST13189995_HCAHeartST13188800     9365\n",
       "HCAHeartST13180619_HCAHeartST13177116     9001\n",
       "HCAHeartST13168796_HCAHeartST13167740     8619\n",
       "HCAHeartST13146207_HCAHeartST13129140     7593\n",
       "HCAHeartST13146204_HCAHeartST13129137     6206\n",
       "HCAHeartST13146205_HCAHeartST13129138     5833\n",
       "HCAHeartST13146206_HCAHeartST13129139     4770\n",
       "HCAHeartST13146208_HCAHeartST13129141     4392\n",
       "HCAHeartST13146202_HCAHeartST13129135     4286\n",
       "HCAHeartST13146203_HCAHeartST13129136     3515\n",
       "HCAHeartST13146201_HCAHeartST13129134     1046\n",
       "Name: combinedID, dtype: int64"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['combinedID'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "welsh-spirituality",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "markdown",
   "id": "efficient-occasion",
   "metadata": {},
   "source": [
    "## Run scrublet"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "bronze-power",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13146201_HCAHeartST13129134\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13146202_HCAHeartST13129135\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13146203_HCAHeartST13129136\n",
      "HCAHeartST13146204_HCAHeartST13129137\n",
      "HCAHeartST13146205_HCAHeartST13129138\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13146206_HCAHeartST13129139\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13146207_HCAHeartST13129140\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13146208_HCAHeartST13129141\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13168796_HCAHeartST13167740\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13189995_HCAHeartST13188800\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13189996_HCAHeartST13188801\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13189997_HCAHeartST13188802\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13180615_HCAHeartST13177112\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13180616_HCAHeartST13177113\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13180617_HCAHeartST13177114\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13180618_HCAHeartST13177115\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "HCAHeartST13180619_HCAHeartST13177116\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/opt/conda/envs/teichlab/lib/python3.8/site-packages/scrublet/helper_functions.py:321: RuntimeWarning: divide by zero encountered in true_divide\n",
      "  w.setdiag(float(target_total) / tots_use)\n",
      "WARNING: Some cells have total count of genes equal to zero\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "CPU times: user 32min 23s, sys: 16min 30s, total: 48min 54s\n",
      "Wall time: 12min 21s\n"
     ]
    }
   ],
   "source": [
    "%%time\n",
    "\n",
    "# per rxn\n",
    "for i,ID in enumerate(adata.obs['combinedID'].unique()):\n",
    "    print(ID)\n",
    "    \n",
    "    ad = adata[adata.obs['combinedID'] == ID].copy()\n",
    "    run_scrublet(ad)\n",
    "    if i==0:\n",
    "        meta = ad.obs\n",
    "    else:\n",
    "        meta = pd.concat([meta, ad.obs])\n",
    "    del ad"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "compact-attack",
   "metadata": {},
   "source": [
    "## Add scrublet outputs to adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "above-muslim",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "(146652, 5)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "meta_scrub = meta[['scrublet_score', 'scrublet_leiden', 'cluster_scrublet_score', 'doublet_pval', 'doublet_bh_pval']].copy()\n",
    "meta_scrub.shape"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "intended-syntax",
   "metadata": {},
   "outputs": [],
   "source": [
    "if meta_scrub.reindex(adata.obs.index).index.equals(adata.obs.index):\n",
    "    adata.obs = pd.concat([adata.obs, meta_scrub.reindex(adata.obs.index)], axis=1)\n",
    "else:\n",
    "    raise Exception('Different barcodes in meta and adata')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "domestic-circuit",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>latent_RT_efficiency</th>\n",
       "      <th>latent_cell_probability</th>\n",
       "      <th>latent_scale</th>\n",
       "      <th>sangerID</th>\n",
       "      <th>combinedID</th>\n",
       "      <th>donor</th>\n",
       "      <th>donor_type</th>\n",
       "      <th>region</th>\n",
       "      <th>region_finest</th>\n",
       "      <th>age</th>\n",
       "      <th>...</th>\n",
       "      <th>facility</th>\n",
       "      <th>cell_or_nuclei</th>\n",
       "      <th>modality</th>\n",
       "      <th>kit_10x</th>\n",
       "      <th>flushed</th>\n",
       "      <th>scrublet_score</th>\n",
       "      <th>scrublet_leiden</th>\n",
       "      <th>cluster_scrublet_score</th>\n",
       "      <th>doublet_pval</th>\n",
       "      <th>doublet_bh_pval</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>barcode</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>HCAHeartST13146201_HCAHeartST13129134_GCAAACCGTCAAGTGC-1</th>\n",
       "      <td>3.620071</td>\n",
       "      <td>1.0</td>\n",
       "      <td>24020.279297</td>\n",
       "      <td>HCAHeartST13146201</td>\n",
       "      <td>HCAHeartST13146201_HCAHeartST13129134</td>\n",
       "      <td>AV10</td>\n",
       "      <td>DBD</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>20-25</td>\n",
       "      <td>...</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.225806</td>\n",
       "      <td>6,3</td>\n",
       "      <td>0.254438</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGTTAGGCGT-1</th>\n",
       "      <td>3.613875</td>\n",
       "      <td>1.0</td>\n",
       "      <td>23438.445312</td>\n",
       "      <td>HCAHeartST13146201</td>\n",
       "      <td>HCAHeartST13146201_HCAHeartST13129134</td>\n",
       "      <td>AV10</td>\n",
       "      <td>DBD</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>20-25</td>\n",
       "      <td>...</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.136531</td>\n",
       "      <td>6,3</td>\n",
       "      <td>0.254438</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCAAAGCTCC-1</th>\n",
       "      <td>3.296565</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15958.253906</td>\n",
       "      <td>HCAHeartST13146201</td>\n",
       "      <td>HCAHeartST13146201_HCAHeartST13129134</td>\n",
       "      <td>AV10</td>\n",
       "      <td>DBD</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>20-25</td>\n",
       "      <td>...</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.254438</td>\n",
       "      <td>6,2</td>\n",
       "      <td>0.254438</td>\n",
       "      <td>0.000043</td>\n",
       "      <td>0.000586</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAGCATGAAG-1</th>\n",
       "      <td>3.371856</td>\n",
       "      <td>1.0</td>\n",
       "      <td>18051.541016</td>\n",
       "      <td>HCAHeartST13146201</td>\n",
       "      <td>HCAHeartST13146201_HCAHeartST13129134</td>\n",
       "      <td>AV10</td>\n",
       "      <td>DBD</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>20-25</td>\n",
       "      <td>...</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.289474</td>\n",
       "      <td>7,0</td>\n",
       "      <td>0.225806</td>\n",
       "      <td>0.000674</td>\n",
       "      <td>0.006062</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>HCAHeartST13146201_HCAHeartST13129134_TGACTCCTCGTAATCA-1</th>\n",
       "      <td>3.205887</td>\n",
       "      <td>1.0</td>\n",
       "      <td>15712.057617</td>\n",
       "      <td>HCAHeartST13146201</td>\n",
       "      <td>HCAHeartST13146201_HCAHeartST13129134</td>\n",
       "      <td>AV10</td>\n",
       "      <td>DBD</td>\n",
       "      <td>SAN</td>\n",
       "      <td>na</td>\n",
       "      <td>20-25</td>\n",
       "      <td>...</td>\n",
       "      <td>Sanger</td>\n",
       "      <td>Nuclei</td>\n",
       "      <td>Multiome-RNA</td>\n",
       "      <td>Multiome-v1</td>\n",
       "      <td>yes</td>\n",
       "      <td>0.254438</td>\n",
       "      <td>7,2</td>\n",
       "      <td>0.213888</td>\n",
       "      <td>0.001828</td>\n",
       "      <td>0.014099</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>5 rows × 21 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    latent_RT_efficiency  \\\n",
       "barcode                                                                    \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...              3.620071   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...              3.613875   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...              3.296565   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...              3.371856   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...              3.205887   \n",
       "\n",
       "                                                    latent_cell_probability  \\\n",
       "barcode                                                                       \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...                      1.0   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...                      1.0   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...                      1.0   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...                      1.0   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...                      1.0   \n",
       "\n",
       "                                                    latent_scale  \\\n",
       "barcode                                                            \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...  24020.279297   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...  23438.445312   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...  15958.253906   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...  18051.541016   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...  15712.057617   \n",
       "\n",
       "                                                              sangerID  \\\n",
       "barcode                                                                  \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...  HCAHeartST13146201   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...  HCAHeartST13146201   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...  HCAHeartST13146201   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...  HCAHeartST13146201   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...  HCAHeartST13146201   \n",
       "\n",
       "                                                                               combinedID  \\\n",
       "barcode                                                                                     \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...  HCAHeartST13146201_HCAHeartST13129134   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...  HCAHeartST13146201_HCAHeartST13129134   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...  HCAHeartST13146201_HCAHeartST13129134   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...  HCAHeartST13146201_HCAHeartST13129134   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...  HCAHeartST13146201_HCAHeartST13129134   \n",
       "\n",
       "                                                   donor donor_type region  \\\n",
       "barcode                                                                      \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...  AV10        DBD    SAN   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...  AV10        DBD    SAN   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...  AV10        DBD    SAN   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...  AV10        DBD    SAN   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...  AV10        DBD    SAN   \n",
       "\n",
       "                                                   region_finest    age  ...  \\\n",
       "barcode                                                                  ...   \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...            na  20-25  ...   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...            na  20-25  ...   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...            na  20-25  ...   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...            na  20-25  ...   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...            na  20-25  ...   \n",
       "\n",
       "                                                   facility cell_or_nuclei  \\\n",
       "barcode                                                                      \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...   Sanger         Nuclei   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...   Sanger         Nuclei   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...   Sanger         Nuclei   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...   Sanger         Nuclei   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...   Sanger         Nuclei   \n",
       "\n",
       "                                                        modality      kit_10x  \\\n",
       "barcode                                                                         \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...  Multiome-RNA  Multiome-v1   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...  Multiome-RNA  Multiome-v1   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...  Multiome-RNA  Multiome-v1   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...  Multiome-RNA  Multiome-v1   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...  Multiome-RNA  Multiome-v1   \n",
       "\n",
       "                                                   flushed scrublet_score  \\\n",
       "barcode                                                                     \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...     yes       0.225806   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...     yes       0.136531   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...     yes       0.254438   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...     yes       0.289474   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...     yes       0.254438   \n",
       "\n",
       "                                                    scrublet_leiden  \\\n",
       "barcode                                                               \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...              6,3   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...              6,3   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...              6,2   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...              7,0   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...              7,2   \n",
       "\n",
       "                                                   cluster_scrublet_score  \\\n",
       "barcode                                                                     \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...               0.254438   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...               0.254438   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...               0.254438   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...               0.225806   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...               0.213888   \n",
       "\n",
       "                                                    doublet_pval  \\\n",
       "barcode                                                            \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...      0.000043   \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...      0.000043   \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...      0.000043   \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...      0.000674   \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...      0.001828   \n",
       "\n",
       "                                                    doublet_bh_pval  \n",
       "barcode                                                              \n",
       "HCAHeartST13146201_HCAHeartST13129134_GCAAACCGT...         0.000586  \n",
       "HCAHeartST13146201_HCAHeartST13129134_ACAGCCGGT...         0.000586  \n",
       "HCAHeartST13146201_HCAHeartST13129134_AGTGTGGCA...         0.000586  \n",
       "HCAHeartST13146201_HCAHeartST13129134_CAAGTGAAG...         0.006062  \n",
       "HCAHeartST13146201_HCAHeartST13129134_TGACTCCTC...         0.014099  \n",
       "\n",
       "[5 rows x 21 columns]"
      ]
     },
     "execution_count": 13,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "advised-madison",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "... storing 'sangerID' as categorical\n",
      "... storing 'combinedID' as categorical\n",
      "... storing 'donor' as categorical\n",
      "... storing 'donor_type' as categorical\n",
      "... storing 'region' as categorical\n",
      "... storing 'region_finest' as categorical\n",
      "... storing 'age' as categorical\n",
      "... storing 'gender' as categorical\n",
      "... storing 'facility' as categorical\n",
      "... storing 'cell_or_nuclei' as categorical\n",
      "... storing 'modality' as categorical\n",
      "... storing 'kit_10x' as categorical\n",
      "... storing 'flushed' as categorical\n",
      "... storing 'scrublet_leiden' as categorical\n",
      "... storing 'gene_name' as categorical\n",
      "... storing 'feature_type' as categorical\n"
     ]
    }
   ],
   "source": [
    "adata.write('/nfs/team205/heart/anndata_objects/8regions/QC/multiome_RNA_adult_new-SAN-AVN_prefilter.h5ad')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "broad-foundation",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 146652 × 36601\n",
       "    obs: 'latent_RT_efficiency', 'latent_cell_probability', 'latent_scale', 'sangerID', 'combinedID', 'donor', 'donor_type', 'region', 'region_finest', 'age', 'gender', 'facility', 'cell_or_nuclei', 'modality', 'kit_10x', 'flushed', 'scrublet_score', 'scrublet_leiden', 'cluster_scrublet_score', 'doublet_pval', 'doublet_bh_pval'\n",
       "    var: 'gene_name', 'feature_type', 'ambient_expression-0', 'ambient_expression-1', 'ambient_expression-10', 'ambient_expression-11', 'ambient_expression-12', 'ambient_expression-13', 'ambient_expression-14', 'ambient_expression-15', 'ambient_expression-16', 'ambient_expression-2', 'ambient_expression-3', 'ambient_expression-4', 'ambient_expression-5', 'ambient_expression-6', 'ambient_expression-7', 'ambient_expression-8', 'ambient_expression-9'\n",
       "    obsm: 'latent_gene_encoding'"
      ]
     },
     "execution_count": 15,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "eight-ferry",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AH2     64903\n",
       "AV13    35489\n",
       "AV14    16809\n",
       "AV3     11985\n",
       "AV10     8847\n",
       "AH1      8619\n",
       "Name: donor, dtype: int64"
      ]
     },
     "execution_count": 16,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata.obs['donor'].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "possible-jersey",
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "median-discovery",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "teichlab",
   "language": "python",
   "name": "teichlab"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
